helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -595,6 +595,14 @@ public RouterUpdater(String clusterName, PropertyType sourceDataType) {
     @Override
     protected void handleEvent(ClusterEvent event) {
       NotificationContext changeContext = event.getAttribute(AttributeName.changeContext.name());
```

Comment:
You shall call setClusterEventId in handleEvent so this information becomes available for the processing methods later.

Reply:
Good point. This is why `BasicClusterDataCache implements ControlContextProvider`.